File: quant1x/config/trader_parameter.go
package config

import (
	"os"
	"sync"

	"github.com/quant1x/quant1x/quant1x/core"
	"gopkg.in/yaml.v3"
)

// TraderParameter mirrors C++ config::TraderParameter (fields only).
type TraderParameter struct {
	AccountId                   string              `yaml:"account_id" json:"account_id"`
	OrderPath                   string              `yaml:"order_path" json:"order_path"`
	TopN                        int                 `yaml:"top_n" json:"top_n"`
	HaveETF                     bool                `yaml:"have_etf" json:"have_etf"`
	PriceCageRatio              float64             `yaml:"price_cage_ratio" json:"price_cage_ratio"`
	MinimumPriceFluctuationUnit float64             `yaml:"minimum_price_fluctuation_unit" json:"minimum_price_fluctuation_unit"`
	FixedSlippageForSell        float64             `yaml:"fixed_slippage_for_sell" json:"fixed_slippage_for_sell"`
	AnnualInterestRate          float64             `yaml:"annual_interest_rate" json:"annual_interest_rate"`
	StampDutyRateForBuy         float64             `yaml:"stamp_duty_rate_for_buy" json:"stamp_duty_rate_for_buy"`
	StampDutyRateForSell        float64             `yaml:"stamp_duty_rate_for_sell" json:"stamp_duty_rate_for_sell"`
	TransferRate                float64             `yaml:"transfer_rate" json:"transfer_rate"`
	CommissionRate              float64             `yaml:"commission_rate" json:"commission_rate"`
	CommissionMin               float64             `yaml:"commission_min" json:"commission_min"`
	PositionRatio               float64             `yaml:"position_ratio" json:"position_ratio"`
	KeepCash                    float64             `yaml:"keep_cash" json:"keep_cash"`
	BuyAmountMax                float64             `yaml:"buy_amount_max" json:"buy_amount_max"`
	BuyAmountMin                float64             `yaml:"buy_amount_min" json:"buy_amount_min"`
	Role                        string              `yaml:"role" json:"role"`
	ProxyUrl                    string              `yaml:"proxy_url" json:"proxy_url"`
	Strategies                  []StrategyParameter `yaml:"strategies" json:"strategies"`
	CancelSession               TradingSession      `yaml:"cancel" json:"cancel"`
	UndertakeRatio              float64             `yaml:"undertake_ratio" json:"undertake_ratio"`
}

var globalTraderOnce sync.Once
var globalTrader *TraderParameter

// loadTraderConfigFromYAML reads trader configuration from a yaml file.
func loadTraderConfigFromYAML(filename string) TraderParameter {
	// initialize with defaults matching C++
	tp := defaultTraderParameter()
	data, err := os.ReadFile(filename)
	if err != nil {
		return tp
	}

	var root yaml.Node
	if err := yaml.Unmarshal(data, &root); err != nil {
		return tp
	}

	// find top-level mapping node
	if len(root.Content) == 0 {
		return tp
	}
	doc := root.Content[0]
	if doc.Kind != yaml.MappingNode {
		return tp
	}

	var traderNode *yaml.Node
	for i := 0; i < len(doc.Content); i += 2 {
		key := doc.Content[i]
		val := doc.Content[i+1]
		if key.Value == "trader" {
			traderNode = val
			break
		}
	}
	if traderNode == nil {
		return tp
	}

	// Unmarshal all trader fields into tp (this may set Strategies but we'll replace it below if necessary)
	if b, err := yaml.Marshal(traderNode); err == nil {
		_ = yaml.Unmarshal(b, &tp)
	}

	// Manually parse strategies to ensure strategy defaults are applied per-item
	for i := 0; i < len(traderNode.Content); i += 2 {
		key := traderNode.Content[i]
		val := traderNode.Content[i+1]
		if key.Value != "strategies" {
			continue
		}
		if val.Kind != yaml.SequenceNode {
			continue
		}
		tp.Strategies = nil
		for _, item := range val.Content {
			sp := defaultStrategyParameter()
			if bs, err := yaml.Marshal(item); err == nil {
				_ = yaml.Unmarshal(bs, &sp)
			}
			tp.Strategies = append(tp.Strategies, sp)
		}
	}

	return tp
}

// defaultTraderParameter returns TraderParameter with C++ default values
func defaultTraderParameter() TraderParameter {
	return TraderParameter{
		AccountId:                   "888xxxxxxx",
		OrderPath:                   "",
		TopN:                        3,
		HaveETF:                     false,
		PriceCageRatio:              ValidDeclarationPriceRange,
		MinimumPriceFluctuationUnit: MinimumPriceFluctuationUnit,
		FixedSlippageForSell:        FixedSlippageForSell,
		AnnualInterestRate:          1.65,
		StampDutyRateForBuy:         0.0,
		StampDutyRateForSell:        0.0010,
		TransferRate:                0.0006,
		CommissionRate:              0.00025,
		CommissionMin:               5.0,
		PositionRatio:               0.5,
		KeepCash:                    10000.0,
		BuyAmountMax:                250000.0,
		BuyAmountMin:                1000.0,
		Role:                        "manual",
		ProxyUrl:                    "http://127.0.0.1:18168/qmt",
		Strategies:                  nil,
		UndertakeRatio:              0.8,
	}
}

// TraderConfig returns the global trader parameter (initialized once)
func TraderConfig() *TraderParameter {
	globalTraderOnce.Do(func() {
		// ensure config filename initialized
		_ = ConfigFilename()
		tp := loadTraderConfigFromYAML(core.GetConfigfilePath())
		globalTrader = &tp
	})
	return globalTrader
}

File: quant1x/config/trader_parameter.cpp
#include <quant1x/config/trader_parameter.h>
#include <quant1x/data/meta/timestamp.h>
#include <quant1x/data/meta/calendar.h>
#include <magic_enum/magic_enum.hpp>

namespace quant1x::config {
namespace meta = quant1x::data::meta;

    // 统计标的总数
    int TraderParameter::TotalNumberOfTargets() const {
        int total = 0;
        for (const auto& v : Strategies) {
            total += v.NumberOfTargets();
        }
        return total;
    }

    // 重置仓位占比
    void TraderParameter::ResetPositionRatio() {
        double remainingRatio = 1.00;
        size_t strategyCount = Strategies.size();
        std::vector<StrategyParameter *> unassignedStrategies;

        for (size_t i = 0; i < strategyCount; i++) {
            StrategyParameter *v = &Strategies[i];
            if (!v->BuyEnable()) {
                continue;
            }
            // 校对个股最大资金
            if (v->FeeMax > BuyAmountMax) {
                v->FeeMax = BuyAmountMax;
            }
            // 校对个股最小资金
            if (v->FeeMin < BuyAmountMin) {
                v->FeeMin = BuyAmountMin;
            }
            if (v->Weight > 1.00) {
                v->Weight = 1.00;
            }
            if (v->Weight > 0) {
                remainingRatio -= v->Weight;
            } else {
                unassignedStrategies.push_back(v);
            }
        }

        size_t remainingCount = unassignedStrategies.size();
        if (remainingRatio > 0 && remainingCount > 0) {
            double averageFundPercentage = remainingRatio / remainingCount;
            for (size_t i = 0; i < remainingCount; i++) {
                StrategyParameter *v = unassignedStrategies[i];
                v->Weight = averageFundPercentage;
                if (i + 1 == remainingCount) {
                    v->Weight = remainingRatio;
                }
                remainingRatio -= v->Weight;
            }
        }
    }

    // 计算每日无风险利率
    double TraderParameter::DailyRiskFreeRate(const std::string& date) const {
        (void)date;
        meta::Timestamp ts = date;
        std::string fixedDate = ts.only_date();
        std::string year = fixedDate.substr(0, 4);
        std::string start = year + "-01-01";
        std::string end = year + "-12-31";
        std::vector<std::string> dates = meta::get_date_range(start, end);
        size_t count = dates.size();
        return AnnualInterestRate / count;
    }

    std::optional<StrategyParameter> TraderParameter::GetStrategyParameterByCode(uint64_t strategyCode) const {
        for (auto const & v : Strategies) {
            if (v.Auto && v.Id == strategyCode) {
                return v;
            }
        }
        return std::nullopt;
    }

    std::ostream &operator<<(std::ostream &os, const TraderParameter &parameter) {
        os << "AccountId: " << parameter.AccountId << "\n"
           << "OrderPath: " << parameter.OrderPath << "\n"
           << "TopN: " << parameter.TopN << "\n"
           << "HaveETF: " << parameter.HaveETF << "\n"
           << "PriceCageRatio: " << parameter.PriceCageRatio << "\n"
           << "MinimumPriceFluctuationUnit: " << parameter.MinimumPriceFluctuationUnit << "\n"
           << "AnnualInterestRate: " << parameter.AnnualInterestRate << "\n"
           << "StampDutyRateForBuy: " << parameter.StampDutyRateForBuy << "\n"
           << "StampDutyRateForSell: " << parameter.StampDutyRateForSell << "\n"
           << "TransferRate: " << parameter.TransferRate << "\n"
           << "CommissionRate: " << parameter.CommissionRate << "\n"
           << "CommissionMin: " << parameter.CommissionMin << "\n"
           << "PositionRatio: " << parameter.PositionRatio << "\n"
           << "KeepCash: " << parameter.KeepCash << "\n"
           << "BuyAmountMax: " << parameter.BuyAmountMax << "\n"
           << "BuyAmountMin: " << parameter.BuyAmountMin << "\n"
           << "Role: " << magic_enum::enum_name(parameter.Role) << "\n"
           << "ProxyUrl: " << parameter.ProxyUrl << "\n"
           << "Strategies:[" << "\n";
        for (auto const & v: parameter.Strategies) {
            os << "  " <<v << "\n";
        }
        os << "]" << "\n";
        //<< " CancelSession: " << parameter.CancelSession
        os << " UndertakeRatio: " << parameter.UndertakeRatio << "\n";
        return os;
    }
} // namespace quant1x::config